# Imports

In [30]:
import pandas as pd
import numpy as np
import pickle
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier

import tensorflow as tf

In [2]:
df = pd.read_csv("11-heart.csv")

# Splitting Data

In [15]:
X = df.drop('output',axis='columns')
y= df['output']

In [16]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [17]:
# Normalization and Standardization

scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

scaler2 = StandardScaler()
X_train_scaled2 = scaler2.fit_transform(X_train)
X_test_scaled2 = scaler2.transform(X_test)

**Decision Tree Classifier**

In [18]:
DT = DecisionTreeClassifier()
DT.fit(X_train_scaled,y_train)

y_prediction = DT.predict(X_test_scaled)
tree_train_acc = round(accuracy_score(y_train,DT.predict(X_train_scaled))*100,2)
tree_test_acc = round(accuracy_score(y_test,y_prediction)*100,2)

print("With normalization:")
print('Decision Tree Train Accuracy = ' , tree_train_acc,'%')
print('Decision tree Test Accuracy = ' , tree_test_acc,'%')
print()
print(classification_report(y_test,DT.predict(X_test_scaled), digits= 4))

With normalization:
Decision Tree Train Accuracy =  100.0 %
Decision tree Test Accuracy =  78.69 %

              precision    recall  f1-score   support

           0     0.7353    0.8621    0.7937        29
           1     0.8519    0.7188    0.7797        32

    accuracy                         0.7869        61
   macro avg     0.7936    0.7904    0.7867        61
weighted avg     0.7964    0.7869    0.7863        61



In [19]:
DT2 = DecisionTreeClassifier()
DT2.fit(X_train_scaled2,y_train)

y_prediction2 = DT2.predict(X_test_scaled2)
tree_train_acc2 = round(accuracy_score(y_train,DT2.predict(X_train_scaled2))*100,2)
tree_test_acc2 = round(accuracy_score(y_test,y_prediction2)*100,2)

print("With standardization:")
print('Decision Tree Train Accuracy = ' , tree_train_acc2,'%')
print('Decision tree Test Accuracy = ' , tree_test_acc2,'%')
print()
print(classification_report(y_test,DT2.predict(X_test_scaled2), digits= 4))

With standardization:
Decision Tree Train Accuracy =  100.0 %
Decision tree Test Accuracy =  77.05 %

              precision    recall  f1-score   support

           0     0.7143    0.8621    0.7813        29
           1     0.8462    0.6875    0.7586        32

    accuracy                         0.7705        61
   macro avg     0.7802    0.7748    0.7699        61
weighted avg     0.7835    0.7705    0.7694        61



**K-Nearest Neighbors Classifier**

In [20]:
KNN = KNeighborsClassifier()
KNN.fit(X_train_scaled,y_train)

y_prediction = KNN.predict(X_test_scaled)
tree_train_acc = round(accuracy_score(y_train,KNN.predict(X_train_scaled))*100,2)
tree_test_acc = round(accuracy_score(y_test,y_prediction)*100,2)

print("With normalization:")
print('KNN Train Accuracy = ' , tree_train_acc,'%')
print('KNN Test Accuracy = ' , tree_test_acc,'%')
print()
print(classification_report(y_test,KNN.predict(X_test_scaled), digits= 4))


With normalization:
KNN Train Accuracy =  84.65 %
KNN Test Accuracy =  85.25 %

              precision    recall  f1-score   support

           0     0.7941    0.9310    0.8571        29
           1     0.9259    0.7812    0.8475        32

    accuracy                         0.8525        61
   macro avg     0.8600    0.8561    0.8523        61
weighted avg     0.8633    0.8525    0.8521        61



In [21]:
KNN2 = KNeighborsClassifier()
KNN2.fit(X_train_scaled2,y_train)

y_prediction2 = KNN2.predict(X_test_scaled2)
tree_train_acc2 = round(accuracy_score(y_train,KNN2.predict(X_train_scaled2))*100,2)
tree_test_acc2 = round(accuracy_score(y_test,y_prediction2)*100,2)

print("With standardization:")
print('KNN Train Accuracy = ' , tree_train_acc2,'%')
print('KNN Test Accuracy = ' , tree_test_acc2,'%')
print()
print(classification_report(y_test,KNN2.predict(X_test_scaled2), digits= 4))

With standardization:
KNN Train Accuracy =  86.31 %
KNN Test Accuracy =  91.8 %

              precision    recall  f1-score   support

           0     0.8750    0.9655    0.9180        29
           1     0.9655    0.8750    0.9180        32

    accuracy                         0.9180        61
   macro avg     0.9203    0.9203    0.9180        61
weighted avg     0.9225    0.9180    0.9180        61



**Random Forest**

In [22]:
RF = RandomForestClassifier()
RF.fit(X_train_scaled,y_train)

y_prediction = RF.predict(X_test_scaled)
tree_train_acc = round(accuracy_score(y_train,RF.predict(X_train_scaled))*100,2)
tree_test_acc = round(accuracy_score(y_test,y_prediction)*100,2)

print("With normalization:")
print('Random Forest Train Accuracy = ' , tree_train_acc,'%')
print('Random Forest Test Accuracy = ' , tree_test_acc,'%')
print()
print(classification_report(y_test,RF.predict(X_test_scaled), digits= 4))

With normalization:
Random Forest Train Accuracy =  100.0 %
Random Forest Test Accuracy =  88.52 %

              precision    recall  f1-score   support

           0     0.8667    0.8966    0.8814        29
           1     0.9032    0.8750    0.8889        32

    accuracy                         0.8852        61
   macro avg     0.8849    0.8858    0.8851        61
weighted avg     0.8858    0.8852    0.8853        61



In [23]:
RF2 = RandomForestClassifier()
RF2.fit(X_train_scaled2,y_train)

y_prediction2 = RF2.predict(X_test_scaled2)
tree_train_acc2 = round(accuracy_score(y_train,RF2.predict(X_train_scaled2))*100,2)
tree_test_acc2 = round(accuracy_score(y_test,y_prediction2)*100,2)

print("With standardization:")
print('Random Forest Train Accuracy = ' , tree_train_acc2,'%')
print('Random Forest Test Accuracy = ' , tree_test_acc2,'%')
print()
print(classification_report(y_test,RF2.predict(X_test_scaled2), digits= 4))

With standardization:
Random Forest Train Accuracy =  100.0 %
Random Forest Test Accuracy =  88.52 %

              precision    recall  f1-score   support

           0     0.8667    0.8966    0.8814        29
           1     0.9032    0.8750    0.8889        32

    accuracy                         0.8852        61
   macro avg     0.8849    0.8858    0.8851        61
weighted avg     0.8858    0.8852    0.8853        61



**Logistic Regression**

In [24]:
LR = LogisticRegression()
LR.fit(X_train_scaled,y_train)

y_prediction = LR.predict(X_test_scaled)
tree_train_acc = round(accuracy_score(y_train,LR.predict(X_train_scaled))*100,2)
tree_test_acc = round(accuracy_score(y_test,y_prediction)*100,2)

print("With normalization:")
print('Logistic Regression Train Accuracy = ' , tree_train_acc,'%')
print('Logistic Regression Test Accuracy = ' , tree_test_acc,'%')
print()
print(classification_report(y_test,LR.predict(X_test_scaled), digits= 4))

With normalization:
Logistic Regression Train Accuracy =  85.48 %
Logistic Regression Test Accuracy =  83.61 %

              precision    recall  f1-score   support

           0     0.8276    0.8276    0.8276        29
           1     0.8438    0.8438    0.8438        32

    accuracy                         0.8361        61
   macro avg     0.8357    0.8357    0.8357        61
weighted avg     0.8361    0.8361    0.8361        61



In [25]:
LR2 = LogisticRegression()
LR2.fit(X_train_scaled2,y_train)

y_prediction2 = LR2.predict(X_test_scaled2)
tree_train_acc2 = round(accuracy_score(y_train,LR2.predict(X_train_scaled2))*100,2)
tree_test_acc2 = round(accuracy_score(y_test,y_prediction2)*100,2)

print("With standardization:")
print('Logistic Regression Train Accuracy = ' , tree_train_acc2,'%')
print('Logistic Regression Test Accuracy = ' , tree_test_acc2,'%')
print()
print(classification_report(y_test,LR2.predict(X_test_scaled2), digits= 4))

With standardization:
Logistic Regression Train Accuracy =  85.89 %
Logistic Regression Test Accuracy =  83.61 %

              precision    recall  f1-score   support

           0     0.8065    0.8621    0.8333        29
           1     0.8667    0.8125    0.8387        32

    accuracy                         0.8361        61
   macro avg     0.8366    0.8373    0.8360        61
weighted avg     0.8380    0.8361    0.8362        61



**Neural Network**

In [26]:
ann = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=8, activation='relu'),   # Input layer
    tf.keras.layers.Dense(units=3, activation='relu'),   # Hidden layer
    tf.keras.layers.Dense(units=3, activation='relu'),   # Hidden layer
    tf.keras.layers.Dense(units=1, activation='sigmoid') # Output layer
])
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
ann.fit(X_train_scaled, y_train, batch_size = 16, epochs = 32)

Epoch 1/32
16/16 [==============================] - 1s 1ms/step - loss: 0.6942 - accuracy: 0.5394
Epoch 2/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6935 - accuracy: 0.5477
Epoch 3/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6929 - accuracy: 0.5477
Epoch 4/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6925 - accuracy: 0.5477
Epoch 5/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6920 - accuracy: 0.5477
Epoch 6/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6913 - accuracy: 0.5477
Epoch 7/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6904 - accuracy: 0.5477
Epoch 8/32
16/16 [==============================] - 0s 939us/step - loss: 0.6894 - accuracy: 0.5477
Epoch 9/32
16/16 [==============================] - 0s 937us/step - loss: 0.6882 - accuracy: 0.5477
Epoch 10/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6860 - accuracy: 0.5477
Epoch 11/32
16/

In [27]:
y_pre = ann.predict(X_test_scaled)
y_pred = (y_pre > 0.5)
cm = confusion_matrix(y_test, y_pred)

acc = round(accuracy_score(y_test,y_pred)*100,2)
print('With normalization:')
print('Neural Network Test Accuracy = ' , acc,'%')
print()
print('Confusion matrix:')
print(cm)
print(classification_report(y_test, y_pred, digits= 4))

2/2 [==============================] - 0s 2ms/step
With normalization:
Neural Network Test Accuracy =  85.25 %

Confusion matrix:
[[22  7]
 [ 2 30]]
              precision    recall  f1-score   support

           0     0.9167    0.7586    0.8302        29
           1     0.8108    0.9375    0.8696        32

    accuracy                         0.8525        61
   macro avg     0.8637    0.8481    0.8499        61
weighted avg     0.8611    0.8525    0.8508        61



In [28]:
ann2 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(units=8, activation='relu'),   # Input layer
    tf.keras.layers.Dense(units=3, activation='relu'),   # Hidden layer
    tf.keras.layers.Dense(units=3, activation='relu'),   # Hidden layer
    tf.keras.layers.Dense(units=1, activation='sigmoid') # Output layer
])
ann2.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
ann2.fit(X_train_scaled2, y_train, batch_size = 16, epochs = 32)

Epoch 1/32
16/16 [==============================] - 1s 1ms/step - loss: 0.6950 - accuracy: 0.4896
Epoch 2/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6869 - accuracy: 0.5353
Epoch 3/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6791 - accuracy: 0.5643
Epoch 4/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6716 - accuracy: 0.5560
Epoch 5/32
16/16 [==============================] - 0s 903us/step - loss: 0.6634 - accuracy: 0.5477
Epoch 6/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6548 - accuracy: 0.5477
Epoch 7/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6455 - accuracy: 0.5477
Epoch 8/32
16/16 [==============================] - 0s 936us/step - loss: 0.6357 - accuracy: 0.5477
Epoch 9/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6264 - accuracy: 0.6473
Epoch 10/32
16/16 [==============================] - 0s 1ms/step - loss: 0.6162 - accuracy: 0.7178
Epoch 11/32
16/

In [29]:
y_pre = ann2.predict(X_test_scaled2)
y_pred = (y_pre > 0.5)
cm = confusion_matrix(y_test, y_pred)

acc2 = round(accuracy_score(y_test,y_pred)*100,2)
print('With standardization:')
print('Neural Network Test Accuracy = ' , acc2,'%')
print()
print('Confusion matrix:')
print(cm)
print(classification_report(y_test, y_pred, digits= 4))

2/2 [==============================] - 0s 994us/step
With standardization:
Neural Network Test Accuracy =  86.89 %

Confusion matrix:
[[26  3]
 [ 5 27]]
              precision    recall  f1-score   support

           0     0.8387    0.8966    0.8667        29
           1     0.9000    0.8438    0.8710        32

    accuracy                         0.8689        61
   macro avg     0.8694    0.8702    0.8688        61
weighted avg     0.8709    0.8689    0.8689        61



# Conclusion
The best one is the KNN with standardization with accuracy 91.8% 